### ISSUE TITLE
Adding cudnn.benchmark = False to set_random_seed

### ISSUE THREAD

**Issue by BramVanroy** [OpenNMT/OpenNMT-py#1784]
When using the CuDNN backend, PyTorch documentation [recommends ](https://pytorch.org/docs/stable/notes/randomness.html#cudnn) disabling `benchmark` operation for reproducibility.

Should this be added to the `set_random_seed`, whose goal is (probably?) to ensure reproducibility.

If requested, I can do a PR.

**Comment by francoishernandez** [comment]
Hey @BramVanroy 
Thanks for the suggestion. Did you notice any issue with the current setup?
Last time I checked reproducibility with fixed random seed was already ok.

**Comment by BramVanroy** [comment]
A colleague of mine was having reproducibility issues when running on different systems, and looking at `set_random_seed`, I can only think of the `benchmark` option that is missing. I do not think he has tried to modify the source code and try to see whether it makes a difference, but perhaps the PyTorch recommendation is grounds enough to implement this? It's just a thought, of course, since (IIRC) `benchmark = False` can possibly lead to slightly worse training speed.

**Comment by francoishernandez** [comment]
Actually I think the issue may rely in the "different systems" part. IIRC it can happen that a perfectly reproducible and deterministic setup on some device(s) is not reproducible on some other.
Not sure whether the `cudnn.benchmark` flag may be of any help in such cases.
It would certainly be interesting to test this on your colleague's setup.

**Comment by BramVanroy** [comment]
We currently don't have time to test this, but reading through [this ](https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936) it seems that PyTorch tries to optimize some things depending on your system that could improve run time speeds. Different systems/architectures/CUDA versions might then lead to different results I suspect. That is probably the reason that the documentation [states ](https://pytorch.org/docs/stable/notes/randomness.html#cudnn) that `torch.backends.cudnn.benchmark = False` must be set, too.

**Comment by francoishernandez** [comment]
Makes sense, but what troubles me though is that we never had any issue regarding this, which leads me into thinking the "must" may be a bit overkill...
Best solution is probably to make this an additional flag in `opts` so that users can set it to their preference without fiddling with the code.
